### PR TITLE
implement Ceil builtin function

### PR DESCRIPTION
**Pull request by yuxubinchen**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #1826

**What this PR does / why we need it:**

implement Ceil builtin function

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

**Comment by broccoliSpicy** [comment]
Our floor/round function applies ClickHouse's floor/round semantic, [reference here](https://clickhouse.com/docs/en/sql-reference/functions/rounding-functions/), which extends the typical floor/round function that takes only one parameter, perhaps MatrixOne's users may get confused if our ceil function doesn't have a consistent semantic as floor/round function

File: pkg/builtin/unary/ceil.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
+package unary
+
+import (
+	"fmt"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
```
Comment (jiajunhuang) [inline pkg/builtin/unary/ceil.go:5]: Split stdlib and self contained lib into two chunks?

Comment (yuxubinchen) [reply pkg/builtin/unary/ceil.go:6]: Sorry, this is my mistake, it has been corrected

**Comment by yuxubinchen** [comment]
> Our floor/round function applies ClickHouse's floor/round semantic, [reference here](https://clickhouse.com/docs/en/sql-reference/functions/rounding-functions/), which extends the typical floor/round function that takes only one parameter, perhaps MatrixOne's users may get confused if our ceil function doesn't have a consistent semantic as floor/round function

Excuse me, does it mean that the ceil function should implement the function of the floor function?

**Comment by yuxubinchen** [comment]
> 我们的 floor/round 函数应用了 ClickHouse 的 floor/round 语义，[参考这里](https://clickhouse.com/docs/en/sql-reference/functions/rounding-functions/)，它扩展了典型的仅接受一个参数的 floor/round 函数，如果我们的 ceil 函数没有与 floor/round 一致的语义，也许 MatrixOne 的用户会感到困惑功能

I don't understand it very well, is it that the Ceil function should implement the function of one parameter of the floor function?

**Comment by yuxubinchen** [comment]
> 我们的 floor/round 函数应用了 ClickHouse 的 floor/round 语义，[参考这里](https://clickhouse.com/docs/en/sql-reference/functions/rounding-functions/)，它扩展了典型的仅接受一个参数的 floor/round 函数，如果我们的 ceil 函数没有与 floor/round 一致的语义，也许 MatrixOne 的用户会感到困惑功能

I don't understand it very well, is it that the Ceil function should implement the function of one parameter of the floor function?

**Comment by broccoliSpicy** [comment]
the floor/round function that MatrixOne already has takes one or two parameters, and behave accordingly, so in my arrogant opinion, it would be better if our ceil function has a accordant behavior with floor/round

perhaps like this
ceil(1.23)    -----> 1.0
ceil(1.23, 0) ----> 1.0
ceil(1.23, 1) -----> 1.3
ceil(12.34, -1) ----> 20.0

File: pkg/builtin/multi/ceil.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+		{
+			Min:        1,
+			Max:        2,
+			Typ:        types.T_uint8,
+			ReturnType: types.T_uint8,
+			Fn: func(vecs []*vector.Vector, proc *process.Process, cs []bool) (*vector.Vector, error) {
+				digits := int64(0)
+				vs := vecs[0].Col.([]uint8)
+				if len(vecs) > 1 {
+					if !cs[1] && vecs[1].Typ.Oid != types.T_int64 {
```
Comment (broccoliSpicy) [inline pkg/builtin/multi/ceil.go:59]: if the second argument is not a constant **or** the second argument's type is not int64, we should both return error, sorry I made this mistake in floor/round function too...

File: pkg/vectorize/ceil/ceil.go [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+func CeilInt32(xs, rs []int32, digits int64) []int32 {
+	return ceilInt32(xs, rs, digits)
+}
+
+func ceilInt32Pure(xs, rs []int32, digits int64) []int32 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits > -MaxUint16digits:
```
Comment (broccoliSpicy) [inline pkg/vectorize/ceil/ceil.go:277]: perhaps this is digits > -MaxInt32digits

File: pkg/vectorize/ceil/ceil.go [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+func CeilUint32(xs, rs []uint32, digits int64) []uint32 {
+	return ceilUint32(xs, rs, digits)
+}
+
+func ceilUint32Pure(xs, rs []uint32, digits int64) []uint32 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits > -MaxInt16digits:
```
Comment (broccoliSpicy) [inline pkg/vectorize/ceil/ceil.go:157]: maybe Uint32

File: pkg/vectorize/ceil/ceil.go [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+	digits := int64(0)
+	for value > 0 {
+		value /= 10
+		digits++
+	}
+	return digits
+}
+
+// scaleTable is a lookup array for digits
+var ScaleTable = [...]uint64{
```
Comment (broccoliSpicy) [inline pkg/vectorize/ceil/ceil.go:56]: perhaps import this ScaleTable and Maxdigits from the floor package would also do the job, and makes code more succinct

File: pkg/builtin/multi/ceil.go [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+	for _, item := range argsAndRets {
+		overload.AppendFunctionRets(builtin.Ceil, item.args, item.ret)
+	}
+	extend.MultiReturnTypes[builtin.Ceil] = func(es []extend.Extend) types.T {
+		return getMultiReturnType(builtin.Ceil, es)
+	}
+
+	extend.MultiStrings[builtin.Ceil] = func(es []extend.Extend) string {
+		if len(es) > 1 {
+			return fmt.Sprintf("ceilr(%s, %s)", es[0], es[1])
```
[the remark this answers is not in the document]
Comment (yuxubinchen) [reply pkg/builtin/multi/ceil.go:43]: Sorry, I made many small mistakes in the details, it has been changed

File: pkg/vectorize/ceil/ceil.go (hunk 3 above)
Comment (yuxubinchen) [reply pkg/vectorize/ceil/ceil.go:277]: problem solved

File: pkg/vectorize/ceil/ceil.go (hunk 4 above)
Comment (yuxubinchen) [reply pkg/vectorize/ceil/ceil.go:157]: problem solved

File: pkg/vectorize/ceil/ceil.go (hunk 5 above)
Comment (yuxubinchen) [reply pkg/vectorize/ceil/ceil.go:56]: problem solved

File: pkg/builtin/multi/ceil.go (hunk 2 above)
Comment (yuxubinchen) [reply pkg/builtin/multi/ceil.go:59]: problem solved

File: pkg/builtin/multi/ceil.go [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+		{
+			Min:        1,
+			Max:        2,
+			Typ:        types.T_uint16,
+			ReturnType: types.T_uint16,
+			Fn: func(vecs []*vector.Vector, proc *process.Process, cs []bool) (*vector.Vector, error) {
+				digits := int64(0)
+				vs := vecs[0].Col.([]uint16)
+				if len(vecs) > 1 {
+					if !cs[1] && vecs[1].Typ.Oid != types.T_int64 {
```
Comment (broccoliSpicy) [inline pkg/builtin/multi/ceil.go:90]: perhaps we need to change here too, the second argument is not a constant **or** the second argument type is not int64..., I think the other lines below have the same mistake

**Review by broccoliSpicy — CHANGES_REQUESTED** [review]

File: pkg/builtin/multi/ceil.go (hunk 7 above)
Comment (yuxubinchen) [reply pkg/builtin/multi/ceil.go:90]: Sorry, I have made changes, the server of the database cannot be connected on my local machine, so I cannot debug effectively

Comment (broccoliSpicy) [reply pkg/builtin/multi/ceil.go]: actually, in matrixone, the constant integer passed in from the frontend will be casted to a int64, so I think the original check was correct
```go
vecs[1].Typ.Oid != types.T_int64
```
what I believe is incorrect here is the logical operation && here, we should return error if any checks here are not satisfied, hence || 
```go
!cs[1] || vecs[1].Typ.Oid != types.T_int64
```

File: pkg/builtin/multi/ceil.go [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+		{
+			Min:        1,
+			Max:        2,
+			Typ:        types.T_uint8,
+			ReturnType: types.T_uint8,
+			Fn: func(vecs []*vector.Vector, proc *process.Process, cs []bool) (*vector.Vector, error) {
+				digits := int64(0)
+				vs := vecs[0].Col.([]uint8)
+				if len(vecs) > 1 {
+					if !cs[1] || vecs[1].Typ.Oid != types.T_uint8 {
```
Comment (broccoliSpicy) [inline pkg/builtin/multi/ceil.go:59]: the constant integer passed in from the frontend will be casted to a int64, so the vector[1] should be a constant int64 vector

**Review by broccoliSpicy — CHANGES_REQUESTED** [review]

File: pkg/builtin/multi/ceil.go (hunk 7 above)
Comment (broccoliSpicy) [reply pkg/builtin/multi/ceil.go]: > Sorry, I have made changes, the server of the database cannot be connected on my local machine, so I cannot debug effectively

you can connect to MatrixOne via a mysql client, after setting up mo-server, run this command in the shell:
```
mysql -h 127.0.0.1 -P 6001 -udump -p
```
or use docker, the project README has the details

File: pkg/builtin/multi/ceil.go (hunk 8 above)
Comment (yuxubinchen) [reply pkg/builtin/multi/ceil.go]: ok,i get your point

**Review by broccoliSpicy — APPROVED** [review]

**Pull request merged**